File: RobotState.lua
--- LUA wrapper for moveit robot state environment
-- dependency to tourch.ros
-- @classmod RobotState
local ffi = require 'ffi'
local torch = require 'torch'
local ros = require 'ros'
local moveit = require 'moveit.env'
local utils = require 'moveit.utils'
local std = ros.std
local tf = ros.tf

local RobotState = torch.class('moveit.RobotState', moveit)

local f

function init()
  local RobotState_method_names = {
    "createEmpty",
    "createFromModel",
    "clone",
    "delete",
    "release",
    "getVariableCount",
    "getVariableNames",
    "getVariablePositions",
    "hasVelocities",
    "getVariableVelocities",
    "hasAccelerations",
    "getVariableAccelerations",
    "hasEffort",
    "getVariableEffort",
    "setToDefaultValues",
    "setToRandomPositions",
    "setFromIK",
    "getGlobalLinkTransform",
    "setVariablePositions",
    "setVariablePositions_",
    "setVariableVelocities",
    "setVariableVelocities_",
    "setVariableAccelerations",
    "setVariableAccelerations_",
    "setVariableEffort",
    "setVariableEffort_",
    "updateLinkTransforms",
    "update",
    "toRobotStateMsg",
    "fromRobotStateMsg",
    "getJointTransform",
    "getJacobian",
    "enforceBounds",
    "distance",
    "satisfiesBounds",
    "copyJointGroupPositions"
  }

  f = utils.create_method_table('moveit_RobotState_', RobotState_method_names)
end

init()

function RobotState.createEmpty()
  local c = torch.factory('moveit.RobotState')()
  rawset(c, 'o', f.createEmpty())
  return c
end

function RobotState.createFromModel(kinematic_model)
  if ffi.typeof(kinematic_model:cdata()) ~= ffi.typeof('RobotModelPtr*') then
    error('RobotState object can only be initialized with existing RobotModel pointer.')
  end
  local c = torch.factory('moveit.RobotState')()
  rawset(c, 'o', f.createFromModel(kinematic_model:cdata()))
  return c
end

function RobotState:__init(o)
  if type(o) ~= 'cdata' or ffi.typeof(o) ~= ffi.typeof('RobotStatePtr*') then
    error('RobotState object can only be initialized with existing RobotState pointer.')
  end
  self.o = o
  ffi.gc(o, f.delete)
  self.moveit_msgs_RobotStateSpec = ros.MsgSpec('moveit_msgs/RobotState')
end

function RobotState:cdata()
  return self.o
end

function RobotState:clone()
  local c = torch.factory('moveit.RobotState')()
  rawset(c, 'o', f.clone(self.o))
  return c
end

---Get the number of variables that make up this state.
--@treturn int
function RobotState:getVariableCount()
  return f.getVariableCount(self.o)
end

---Get the names of the variables that make up this state, in the order they are stored in memory.
--@tparam[opt] moveit.Strings names
--@treturn moveit.Strings
function RobotState:getVariableNames(names)
  names = names or std.StringVector()
  f.getVariableNames(self.o, names:cdata())
  return names
end

---Get a raw pointer to the positions of the variables stored in this state. Use carefully.
--If you change these values externally you need to make sure you trigger a forced update for the state by calling update(true).
--@treturn torch.DoubleTensor
function RobotState:getVariablePositions()
  local t = torch.DoubleTensor()
  f.getVariablePositions(self.o, t:cdata())
  return t
end

---By default, if velocities are never set or initialized, the state remembers that there are no velocities set.
--This is useful to know when serializing or copying the state.
--@treturn boolean
function RobotState:hasVelocities()
  return f.hasVelocities(self.o)
end

---Get const access to the velocities of the variables that make up this state.
--The values are in the same order as reported by getVariableNames.
--@see getVariableNames
--@treturn torch.DoubleTensor
function RobotState:getVariableVelocities()
  local t = torch.DoubleTensor()
  f.getVariableVelocities(self.o, t:cdata())
  return t
end

---Check if accelerations are provided.
--By default, if accelerations are never set or initialized, the state remembers that there are no accelerations set.
--This is useful to know when serializing or copying the state. If hasAccelerations() reports true, hasEffort() will certainly report false.
--@treturn boolean
function RobotState:hasAccelerations()
  return f.hasAccelerations(self.o)
end

---Get raw access to the accelerations of the variables that make up this state.
--The values are in the same order as reported by getVariableNames().
--The area of memory overlaps with effort (effort and acceleration should not be set at the same time)
--@treturn torch.DoubleTensor
--@see getVariableNames
function RobotState:getVariableAccelerations()
  local t = torch.DoubleTensor()
  f.getVariableAccelerations(self.o, t:cdata())
  return t
end

---Check if efforts are provided.
--By default, if effort is never set or initialized, the state remembers that there is no effort set.
--This is useful to know when serializing or copying the state. If hasEffort() reports true, hasAccelerations() will certainly report false.
--@treturn boolean
function RobotState:hasEffort()
  return f.hasEffort(self.o)
end

---Get raw access to the effort of the variables that make up this state.
--The values are in the same order as reported by getVariableNames().
--@treturn torch.DoubleTensor
--@see getVariableNames
function RobotState:getVariableEffort()
  local t = torch.DoubleTensor()
  f.getVariableEffort(self.o, t:cdata())
  return t
end

---Set all joints to their default positions.
--The default position is 0, or if that is not within bounds then half way between min and max bound.
function RobotState:setToDefaultValues()
  f.setToDefaultValues(self.o)
end

---Set all joints to random values.
--Values will be within default bounds.
function RobotState:setToRandomPositions()
  f.setToRandomPositions(self.o)
end

---If the group this state corresponds to is a chain and a solver is available, then the joint values can be set by computing inverse kinematics.
--The pose is assumed to be in the reference frame of the kinematic model.
--Returns true on success.
--@tparam string group_id
--@tparam tf.Transform pose
--@tparam[opt=10] int attempts
--@tparam[opt=0.1] number timeout
--@treturn boolean
function RobotState:setFromIK(group_id, pose, attempts, timeout, return_approximate_solution)
  attempts = attempts or 10
  timeout = timeout or 0.1
  return_approximate_solution = return_approximate_solution or false
  if torch.isTensor(pose) then
      pose = tf.Transform():fromTensor(pose)
  end
  local result_joint_positions = torch.DoubleTensor()
  local found_ik =
    f.setFromIK(
    self.o,
    group_id,
    pose:cdata(),
    attempts,
    timeout,
    return_approximate_solution,
    result_joint_positions:cdata()
  )
  return found_ik, result_joint_positions
end

---Do forward Kinematic
--@tparam tf.Transform pose
--@tparam string link name
function RobotState:getGlobalLinkTransform(link_name, pose)
  pose = pose or tf.Transform()
  f.getGlobalLinkTransform(self.o, pose:cdata(), link_name)
  return pose
end

---It is assumed positions is an array containing the new positions for all variables in this state. Those values are copied into the state.
--
function RobotState:setVariablePositions(group_variable_values, group_variable_names)
    if group_variable_names then
        if torch.type(group_variable_names) == 'table' then
            group_variable_names = std.StringVector(group_variable_names)
        end
        if torch.type(group_variable_names) == 'std.StringVector' then
            return f.setVariablePositions_(self.o, group_variable_values:cdata(), group_variable_names:cdata())
        else
            ros.ERROR('[RobotState:setVariablePositions] Could not set Positions ini robot state')
        end
    else
        return f.setVariablePositions(self.o, group_variable_values:cdata())
    end
end

---It is assumed positions is an array containing the new velocities for all variables in this state. Those values are copied into the state.
--
function RobotState:setVariableVelocities(group_variable_values)
  if group_variable_names then
    if torch.type(group_variable_names) == 'table' then
      group_variable_names = std.StringVector(group_variable_names)
    end
    if torch.type(group_variable_names) == 'std.StringVector' then
      return f.setVariableVelocities_(self.o, group_variable_values:cdata(), group_variable_names:cdata())
    else
      ros.ERROR('[RobotState:setVariableVelocities] Could not set Positions ini robot state')
    end
  else
    return f.setVariableVelocities(self.o, group_variable_values:cdata())
  end
end

---It is assumed positions is an array containing the new accelerations for all variables in this state. Those values are copied into the state.
--
function RobotState:setVariableAccelerations(group_variable_values)
  if group_variable_names then
    if torch.type(group_variable_names) == 'table' then
        group_variable_names = std.StringVector(group_variable_names)
    end
    if torch.type(group_variable_names) == 'std.StringVector' then
        return f.setVariableAccelerations_(self.o, group_variable_values:cdata(), group_variable_names:cdata())
    else
        ros.ERROR('[RobotState:setVariableVelocities] Could not set Positions ini robot state')
    end
  else
    return f.setVariableAccelerations(self.o, group_variable_values:cdata())
  end
end

---It is assumed positions is an array containing the new effort for all variables in this state. Those values are copied into the state.
--
function RobotState:setVariableEffort(group_variable_values)
  return f.setVariableEffort(self.o, group_variable_values:cdata())
end

---Update the reference frame transforms for links. This call is needed before using the transforms of links for coordinate transforms.
--
function RobotState:updateLinkTransforms()
  f.updateLinkTransforms(self.o)
end

---Update all transforms.
--
function RobotState:update()
  f.update(self.o)
end

---Convert Robotstate to Message format
--@tparam boolean copy_attached_bodies
function RobotState:toRobotStateMsg(copy_attached_bodies)
  copy_attached_bodies = copy_attached_bodies or false
  local msg_bytes = torch.ByteStorage()
  f.toRobotStateMsg(self.o, msg_bytes:cdata(), copy_attached_bodies)
  local msg = ros.Message(self.moveit_msgs_RobotStateSpec or ros.MsgSpec('moveit_msgs/RobotState'), true)
  msg:deserialize(msg_bytes)
  return msg
end

---Convert Message format to Robotstate
--@tparam Robotstate_msg msg
function RobotState:fromRobotStateMsg(msg)
  if torch.isTypeOf(msg, ros.Message) then
    local msg_bytes = msg:serialize()
    msg_bytes:shrinkToFit()
    f.fromRobotStateMsg(self.o, msg_bytes.storage:cdata())
    return true
  end
  return false
end

---Determines joint transformation from joint name
--@tparam string joint_name
function RobotState:getJointTransform(joint_name)
  assert(joint_name and type(joint_name) == 'string' and #joint_name >= 0, 'Invalid joint_name specified.')
  local result = torch.DoubleTensor()
  f.getJointTransform(self.o, joint_name, result:cdata())
  return result
end

---Determines Jacobian for move group name
--@tparam string group_name
function RobotState:getJacobian(group_name)
  assert(group_name and type(group_name) == 'string' and #group_name >= 0, 'Invalid group_name specified.')
  local result = torch.DoubleTensor()
  f.getJacobian(self.o, group_name, result:cdata())
  return result
end

function RobotState:enforceBounds()
  f.enforceBounds(self.o)
end

function RobotState:distance(other)
  if torch.isTypeOf(other, moveit.RobotState) then
    return f.distance(self.o, other:cdata())
  end
end

function RobotState:satisfiesBounds(margin)
  local margin = margin or 0.0
  return f.satisfiesBounds(self.o, margin)
end

function RobotState:copyJointGroupPositions(group_name)
  assert(group_name and type(group_name) == 'string' and #group_name >= 0, 'Invalid group_name specified.')
  local t = torch.DoubleTensor()
  f.copyJointGroupPositions(self.o, group_name, t:cdata())
  return t
end

File: src/robot_state.cpp
#include "torch-moveit.h"
#include <moveit/robot_state/conversions.h>
#include "utils.h"

MOVIMP(RobotStatePtr *, RobotState, createEmpty)()
{
  return new RobotStatePtr();
}

MOVIMP(RobotStatePtr *, RobotState, createFromModel)(RobotModelPtr *kinematic_model)
{
  robot_state::RobotStatePtr kinematic_state(new robot_state::RobotState(*kinematic_model));
  kinematic_state->setToDefaultValues();
  return new RobotStatePtr(kinematic_state);
}

MOVIMP(RobotStatePtr *, RobotState, clone)(RobotStatePtr *ptr)
{
  return new RobotStatePtr(new moveit::core::RobotState(**ptr));
}

MOVIMP(void, RobotState, delete)(RobotStatePtr *ptr)
{
  delete ptr;
}

MOVIMP(void, RobotState, release)(RobotStatePtr *ptr)
{
  ptr->reset();
}

MOVIMP(int, RobotState, getVariableCount)(RobotStatePtr *self)
{
  return static_cast<int>((*self)->getVariableCount());
}

MOVIMP(void, RobotState, getVariableNames)(RobotStatePtr *self, StringVector *output)
{
  *output = (*self)->getVariableNames();
}

MOVIMP(void, RobotState, getVariablePositions)(RobotStatePtr *self, THDoubleTensor *view)
{
  double *data = (*self)->getVariablePositions();
  size_t count = (*self)->getVariableCount();
  viewArray(data, count, view);
}

MOVIMP(bool, RobotState, hasVelocities)(RobotStatePtr *self)
{
  return (*self)->hasVelocities();
}

MOVIMP(void, RobotState, getVariableVelocities)(RobotStatePtr *self, THDoubleTensor *view)
{
  double *data = (*self)->getVariableVelocities();
  size_t count = (*self)->getVariableCount();
  viewArray(data, count, view);
}

MOVIMP(void, RobotState, setVariableVelocities)(RobotStatePtr *self, THDoubleTensor *view)
{
  std::vector<double> group_variable_values;
  Tensor2vector(view,group_variable_values);
  return (*self)->setVariableVelocities (group_variable_values);
}

MOVIMP(void, RobotState, setVariableVelocities_)(RobotStatePtr *self, THDoubleTensor *t, StringVector *input) {
  std::vector<double> group_variable_values;
  Tensor2vector(t,group_variable_values);
  (*self)->setVariableVelocities(*input, group_variable_values);
}

MOVIMP(bool, RobotState, hasAccelerations)(RobotStatePtr *self)
{
  return (*self)->hasAccelerations();
}

MOVIMP(void, RobotState, getVariableAccelerations)(RobotStatePtr *self, THDoubleTensor *view)
{
  double *data = (*self)->getVariableAccelerations();
  size_t count = (*self)->getVariableCount();
  viewArray(data, count, view);
}

MOVIMP(void, RobotState, setVariableAccelerations)(RobotStatePtr *self, THDoubleTensor *view)
{
  std::vector<double> group_variable_values;
  Tensor2vector(view,group_variable_values);
  return (*self)->setVariableAccelerations (group_variable_values);
}

MOVIMP(void, RobotState, setVariableAccelerations_)(RobotStatePtr *self, THDoubleTensor *t, StringVector *input) {
  std::vector<double> group_variable_values;
  Tensor2vector(t,group_variable_values);
  (*self)->setVariableAccelerations(*input, group_variable_values);
}

MOVIMP(bool, RobotState, hasEffort)(RobotStatePtr *self)
{
  return (*self)->hasEffort();
}

MOVIMP(void, RobotState, getVariableEffort)(RobotStatePtr *self, THDoubleTensor *view)
{
  double *data = (*self)->getVariableEffort();
  size_t count = (*self)->getVariableCount();
  viewArray(data, count, view);
}

MOVIMP(void, RobotState, setVariableEffort)(RobotStatePtr *self, THDoubleTensor *view)
{
  std::vector<double> group_variable_values;
  Tensor2vector(view,group_variable_values);
  return (*self)->setVariableEffort (group_variable_values);
}

MOVIMP(void, RobotState, setVariableEffort_)(RobotStatePtr *self, THDoubleTensor *t, StringVector *input) {
  std::vector<double> group_variable_values;
  Tensor2vector(t,group_variable_values);
  (*self)->setVariableEffort(*input, group_variable_values);
}

MOVIMP(void, RobotState, setToDefaultValues)(RobotStatePtr *self)
{
  (*self)->setToDefaultValues();
}

MOVIMP(void, RobotState, setToRandomPositions)(RobotStatePtr *self)
{
  (*self)->setToRandomPositions();
}

MOVIMP(bool, RobotState, setFromIK)(
  RobotStatePtr *self,
  const char *group_id,
  const tf::Transform *pose,
  const char *tip,
  unsigned int attempts,
  double timeout,
  bool return_approximate_solution,
  THDoubleTensor *result_joint_positions
) {
  const robot_state::JointModelGroup *group = (*self)->getJointModelGroup(group_id);

  std::vector<double> joint_values;
  Eigen::Affine3d pose_;
  tf::poseTFToEigen(*pose, pose_);

  kinematics::KinematicsQueryOptions query_options;
  if (return_approximate_solution) {
    query_options.return_approximate_solution = true;
  }
  std::string tip_link = std::string(tip);
  bool found_ik = false;
  if (tip_link.size() == 0 )
  {
    found_ik = (*self)->setFromIK(group, pose_, attempts, timeout, moveit::core::GroupStateValidityCallbackFn(), query_options);
  } else {
    found_ik = (*self)->setFromIK(group, pose_, tip_link, attempts, timeout, moveit::core::GroupStateValidityCallbackFn(), query_options);
  }


  if (found_ik) {
    (*self)->copyJointGroupPositions(group, joint_values);
    vector2Tensor(joint_values, result_joint_positions);
  } else {
    ROS_DEBUG("Did not find IK solution");
  }

  return found_ik;
}

MOVIMP(bool, RobotState, getGlobalLinkTransform)(
  RobotStatePtr *self,
  tf::Transform *pose_,
  const char *link_name
) {
  const moveit::core::LinkModel* model = (*self)->getLinkModel(link_name);
  if (model == NULL)
  {
    return false;
  }
  Eigen::Affine3d pose = (*self)->getGlobalLinkTransform(model);
  tf::poseEigenToTF(pose, *pose_);
  return true;
}

MOVIMP(void, RobotState, setVariablePositions)(RobotStatePtr *self, THDoubleTensor *t) {
  std::vector<double> group_variable_values;
  Tensor2vector(t,group_variable_values);
  (*self)->setVariablePositions(group_variable_values);
}

MOVIMP(void, RobotState, setVariablePositions_)(RobotStatePtr *self, THDoubleTensor *t, StringVector *input) {
  std::vector<double> group_variable_values;
  Tensor2vector(t,group_variable_values);
  (*self)->setVariablePositions(*input, group_variable_values);
}

MOVIMP(void, RobotState, updateLinkTransforms)(RobotStatePtr *self) {
  (*self)->updateLinkTransforms();
}

MOVIMP(void, RobotState, update)(RobotStatePtr *self) {
  (*self)->update();
}

MOVIMP(void, RobotState, toRobotStateMsg)(RobotStatePtr *self, THByteStorage *output, bool copy_attached_bodies) {
  moveit_msgs::RobotState robot_state;

  robot_state::robotStateToRobotStateMsg(**self, robot_state, copy_attached_bodies);

  uint32_t length = ros::serialization::serializationLength(robot_state);
  THByteStorage_resize(output, length + sizeof(uint32_t));
  ros::serialization::OStream stream(THByteStorage_data(output), length + sizeof(uint32_t));
  stream.next((uint32_t)length);
  ros::serialization::serialize(stream, robot_state);
}

MOVIMP(void, RobotState, fromRobotStateMsg)(RobotStatePtr *self, THByteStorage *serialized_msg)
{
  long storage_size = THByteStorage_size(serialized_msg);

  uint8_t *data = THByteStorage_data(serialized_msg);

  ros::serialization::IStream stream(data+ sizeof(uint32_t), static_cast<uint32_t>(storage_size-sizeof(uint32_t)));
  moveit_msgs::RobotState rs_msg;
  ros::serialization::Serializer<moveit_msgs::RobotState>::read(stream, rs_msg);
  robot_state::robotStateMsgToRobotState(rs_msg, (**self));
}

MOVIMP(void, RobotState, getJointTransform)(RobotStatePtr *self, const char *joint_name, THDoubleTensor *result) {
  const Eigen::Affine3d & pose = (*self)->getJointTransform (joint_name);
  copyMatrix(pose.matrix(), result);
}

MOVIMP(void, RobotState, getJacobian)(RobotStatePtr *self, const char *group_id, THDoubleTensor *result) {
  const robot_state::JointModelGroup *group = (*self)->getJointModelGroup(group_id);
  Eigen::Vector3d linkOrigin = Eigen::Vector3d::Zero();
  Eigen::MatrixXd jacobian; //= (*self)->getJacobian (group,linkOrigin);
  (*self)->getJacobian(group, (*self)->getLinkModel(group->getLinkModelNames().back()),linkOrigin, jacobian);
  THDoubleTensor_resize2d(result, jacobian.rows(), jacobian.cols());
  THDoubleTensor* result_ = THDoubleTensor_newContiguous(result);
  Eigen::Map<Eigen::Matrix<double,Eigen::Dynamic,Eigen::Dynamic,Eigen::RowMajor>> m (THDoubleTensor_data(result_),jacobian.rows(), jacobian.cols());
  m = jacobian;
  THDoubleTensor_freeCopyTo(result_, result);
}

MOVIMP(void, RobotState, enforceBounds)(RobotStatePtr *self) {
  (*self)->enforceBounds();
}

MOVIMP(double, RobotState, distance)(RobotStatePtr *self, RobotStatePtr *other) {
  return (*self)->distance(**other);
}

MOVIMP(bool, RobotState, satisfiesBounds)(RobotStatePtr *self, double margin) {
  return (*self)->satisfiesBounds(margin);
}

MOVIMP(void, RobotState, copyJointGroupPositions)(RobotStatePtr *self, const char *group_id, THDoubleTensor *result) {
  std::vector<double> joint_group_positions;
  (*self)->copyJointGroupPositions(group_id, joint_group_positions);
  vector2Tensor(joint_group_positions,result);
}
/*
void 	printDirtyInfo (std::ostream &out=std::cout) const
void 	printStateInfo (std::ostream &out=std::cout) const
void 	printStatePositions (std::ostream &out=std::cout) const
void 	printTransform (const Eigen::Affine3d &transform, std::ostream &out=std::cout) const
void 	printTransforms (std::ostream &out=std::cout) const
*/
